cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -171,6 +119,78 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
+			bootstrapapi.KubeConfigKey, bootstrapapi.ConfigMapClusterInfo)
+	}
+	detachedJWSToken, ok := insecureClusterInfo.Data[bootstrapapi.JWSSignatureKeyPrefix+token.ID]
+	if !ok || len(detachedJWSToken) == 0 {
+		return nil, nil, errors.Errorf("token id %q is invalid for this cluster or it has expired. Use \"kubeadm token create\" on the control-plane node to create a new valid token", token.ID)
+	}
+	if !bootstrap.DetachedTokenIsValid(detachedJWSToken, insecureKubeconfigString, token.ID, token.Secret) {
+		return nil, nil, errors.New("failed to verify JWS signature of received cluster info object, can't trust this API Server")
+	}
+	insecureKubeconfigBytes := []byte(insecureKubeconfigString)
+	insecureConfig, err := clientcmd.Load(insecureKubeconfigBytes)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "couldn't parse the kubeconfig file in the %s configmap", bootstrapapi.ConfigMapClusterInfo)
+	}
+	return insecureKubeconfigBytes, insecureConfig, nil
+}
+
+// getSecureKubeconfig connects to the API Server and tries to fetch the cluster-info ConfigMap,
+// then validate whether the kubeconfig from the securely-obtained ConfigMap is the same as insecureKubeconfigBytes or not,
+// if they are the same, then Load secureKubeconfigBytes and return, otherwise return error
+func getSecureKubeconfig(secureBootstrapConfig *clientcmdapi.Config, insecureKubeconfigBytes []byte) (*clientcmdapi.Config, error) {
+	secureClusterInfo, err := getClusterInfoFromConfig(secureBootstrapConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pull the kubeconfig from the securely-obtained ConfigMap and validate that it's the same as what we found the first time
+	secureKubeconfigBytes := []byte(secureClusterInfo.Data[bootstrapapi.KubeConfigKey])
+	if !bytes.Equal(secureKubeconfigBytes, insecureKubeconfigBytes) {
+		return nil, errors.Errorf("the second kubeconfig from the %s configmap (using validated TLS) was different from the first", bootstrapapi.ConfigMapClusterInfo)
+	}
+
+	secureKubeconfig, err := clientcmd.Load(secureKubeconfigBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't parse the kubeconfig file in the %s configmap", bootstrapapi.ConfigMapClusterInfo)
+	}
+	return secureKubeconfig, nil
+}
+
+// getClusterInfoFromConfig connects to the API Server and tries to fetch the cluster-info ConfigMap
+func getClusterInfoFromConfig(config *clientcmdapi.Config) (*v1.ConfigMap, error) {
+	client, err := kubeconfigutil.ToClientSet(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var clusterInfo *v1.ConfigMap
+	wait.PollImmediateInfinite(constants.DiscoveryRetryInterval, func() (bool, error) {
+		var err error
+		clusterInfo, err = client.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
+		if err != nil {
+			klog.V(1).Infof("[discovery] Failed to request cluster info, will try again: [%s]\n", err)
+			return false, nil
```

Comment:
did you want to return `err` here instead of nil?

Reply:
no, it will try again.